### ISSUE TITLE
plugin info enhancements

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#2098]
Reading up on:

http://tiddlywiki.com/prerelease/#PluginMechanism

I find a lot of plugin fields that don't appear to be exposed in either of:

* the controlpanel
* the info template for a plugin tiddler

## Improvements

1. most crucially: add a `docs` field to the **plugin.info** template for the author to specify a link to the full plugin documentation
    1. I think we agree that shipping a plugin with all its details and demos does not sound like a practice that should be recommended
    2. the `source` for the plugin is not quite the same as its `docs`
        1. to me `source` should be a link to a place like a github repo holding the source code
2. show the same plugin info when expanding a plugin in the ControlPanel as is shown in the info template when opening the plugin tiddler
    1. to reduce complexity
    2. to do away with a certain confusion as to where to look for what
    3. **all** information contained in a **plugin.info** should be exposed / transcluded by the core ui where appropriate, not just selected bits
3. expose more plugin details as specified via **plugin.info**
    1. most importantly: there needs to be a link to either `source` and/or `docs`
4. the plugin name should be automatically extracted and clearly exposed in the ui:
    1. either as **pluginname** (the bit of the plugin title after the last slash) or **authorname/pluginname**
        2. while that information is shown in the title, it is suboptimal that we don't see it prominently in the plugin list

**Comment by Jermolene** [comment]
Hi @tobibeer I agree with your points, and would welcome these changes. @felixhayashi and others publishing plugins, do you have any thoughts?

**Comment by felixhayashi** [comment]
Hi @Jermolene @tobibeer 

Good idea to focus on the plugin section

I think I agree with @tobibeer's points.

One comments/additions:

> the plugin name should be automatically extracted and clearly exposed in the ui: 

I would go one step further. At the moment, the convention is that the plugin-tiddler title is written `$:/plugins/author-in-small-letters/title-in-small-letters`. However, I consider this the internal "codename" and displaying it would look pretty ugly (e.g. `googleanalytics` vs. `Google Analytics` or `markdown` vs. `Markdown Parser`). Currently we use the `description` field as title but this usually also contains a short intro line and not just the title – the result is, there is no field that holds the mere official plugin title (in contrast to the internal codename or tiddler path). One can argue that it is possible to only write the title in the `description` field but what if I have a slogan or short description for the plugin that should be visible in the plugin list?

Maybe we can add a field that serves this purpose using `description` as fallback. E.g. `plugin-title` (similar to `plugin-type`) or `name`.

It would be great if the plugin section showed all plugins initially like this:

* Title (version)
* Description

When clicking on a plugin It shouldn't open the plugin tiddler but show the plugin infos in a nicely formatted way together (what Tobias said) but also with the "shipped text files" (license, readme) as tabs denoted in the list field. Many people click in the middle of a listed plugin and the plugin tiddler opens itself, I think there should be a link to this tiddler but it is confusing if this tiddler is opened instead of plugin infos (as tobias suggested).

-Felix

**Comment by Jermolene** [comment]
Good points thank you @felixhayashi 

I agree about the problem of the lack of a human readable unique title for plugins. My original plan for plugin names was to detect and specially suppress the `$:/plugins/` prefix, giving us names like "hayashi/tiddlymap`. Such titles would be displayed in a nice big font, while plugins whose title lacked the prefix would be displayed in full, in a smaller grey font. (All very similar to the way that we detect and process the `$:/` system tiddler prefix in the view template for tiddler titles).

The obvious problem with a `plugin-title` field is that we really want it to be unique, but there isn't a mechanism to make it so.

**Comment by tobibeer** [comment]
> The obvious problem with a plugin-title field is that we really want it to be unique, but there isn't a mechanism to make it so.

I don't think we can or should aspire to achieve that. Plugins can have the same readable name, just because authors gave it to them. The uniquely identifying bit is indeed in the plugin name, not necessarily a more descriptive title... although that would surely be desirable.

In any case, using the description alone is not the best solution to list plugins. Also, plugins should be sorted by their full title, imho.

**Comment by felixhayashi** [comment]
> The uniquely identifying bit is indeed in the plugin tiddler name, not necessarily a more descriptive title... although that would surely be desirable.

I agree with @tobibeer here. We could eliminate this a bit by adding the author's name to the plugin list:


\[Title] (\[version]) authored by \[Author]
\[Description]

@Jermolene 

>  My original plan for plugin names was to detect and specially suppress the $:/plugins/ prefix, giving us names like "hayashi/tiddlymap. Such titles would be displayed in a nice big font, while plugins whose title lacked the prefix would be displayed in full, in a smaller grey font. (All very similar to the way that we detect and process the$:/`

But then we only show the codename which will look quite ugly (regardless of the font chosen) and create inner conflicts for plugin creators how to name the plugin because the codename were also the publically visible name and not just a compact abbreviation for internal use. Consider e.g.

| Codename              | Official name                |
|-----------------------|------------------------------|
| googleanalytics       | Google Analytics             |
| powered-by-tiddlywiki | Powered by TiddlyWiki Banner |
| railroad              | Railroad diagram generator   |
| stacked-view          | Stacked card storyview       |

Codenames are cryptic, I would not want users to see them in a list of plugins.

> (All very similar to the way that we detect and process the `$:/` system tiddler prefix in the view template for tiddler titles).

When plugins are listed, they need to be presented in a way that makes them easy to understand just by looking at the title. 

Compare:

 `$:/harrosb/brallo`

**to**

`Brallo visualization library (authored by Harro Sparodine)`

Looking at a string `$:/harrosb/brallo` doesn't say anything but `Brallo visualization library (authored by Harro Sparodine)` makes instant sense.

-Felix

**Comment by tobibeer** [comment]
@felixhayashi, I very much agree that we should pull nicely formatted "pretty info" from the plugin-info, as available, with appropriate fallback handling should those fields not be defined.

**Comment by felixhayashi** [comment]
>  with appropriate fallback handling should those fields not be defined.

Yap

**Comment by pmario** [comment]
- The issue was created with [TW v5.1.9](https://tiddlywiki.com/archive/full/TiddlyWiki-5.1.9) in mind
- **Old plugin overview:** [$:/tags/ControlPanel](https://tiddlywiki.com/archive/full/TiddlyWiki-5.1.9#%24%3A%2Fcore%2Fui%2FControlPanel%2FPlugins)

---

- The plugins have a **name** field now, that is shown as plugin name. 
- The new [Plugins overview](https://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FControlPanel%2FPlugins)
- Plugin details have a **contents** tab now
- Opening the plugin tiddler shows the UI as the control panel overview.
- So the tiddler (i)nfo button is easy to access, which shows the plugin fields. 

All the requirements of the OP are met.
